### PR TITLE
Delete finished jobs

### LINKED ISSUE #344: Delete finished jobs
We could offer the option to automatically delete jobs from the DB as soon as they're finished. Im guessing that for many use cases, this would be much more useful than running a periodic job to delete them later.

(We could also imagine a hooks system where we could provide builtin actions or let users define their custom actions)

### PR DESCRIPTION
**Pull request by mxd4**
Closes #344

This PR adds a new parameter when creating a worker, allowing the deletion of jobs on completion,
either for successful jobs or all jobs.

### Successful PR Checklist:

- [x] Tests
  - [ ] (not applicable?)
- [x] Documentation
  - [ ] (not applicable?)
- [x] Had a good time contributing?

### CODE REVIEW

**Comment by ewjoachim** [comment]
Woo, I'm excited to see the rest. The start is looking awesome !

File: procrastinate/cli.py [hunk 1]
```diff
@@ -167,6 +167,12 @@ def close_connection(procrastinate_app: procrastinate.App, *args, **kwargs):
     default=True,
     help="Whether to actively listen for new jobs or periodically poll",
 )
+@click.option(
+    "--delete-jobs",
+    type=click.Choice([v.value for v in worker.DeleteJobCondition]),
```
Comment (ewjoachim) [inline procrastinate/cli.py:172]: Good job on that ! I'm impressed :)

File: procrastinate/periodic.py [hunk 2]
```diff
@@ -72,7 +72,7 @@ async def worker(self) -> None:
 
     def register_task(self, task: tasks.Task, cron: str) -> PeriodicTask:
         logger.info(
-            f"Registring task {task.name} to run periodically with cron {cron}",
+            f"Registering task {task.name} to run periodically with cron {cron}",
```
Comment (ewjoachim) [inline procrastinate/periodic.py:75]: Haha oops :sweat_smile:

File: tests/integration/test_manager.py [hunk 3]
```diff
@@ -231,15 +235,17 @@ async def test_finish_job(get_all, pg_job_manager, job_factory):
...
     assert await get_all("procrastinate_jobs", "status", "attempts") == expected
 
 
 async def test_finish_job_retry(get_all, pg_job_manager, job_factory):
     await pg_job_manager.defer_job_async(job_factory())
     job1 = await pg_job_manager.fetch_job(queues=None)
-    await pg_job_manager.finish_job(job=job1, status=jobs.Status.TODO)
+    await pg_job_manager.finish_job(job=job1, status=jobs.Status.TODO, delete_job=False)
 
     job2 = await pg_job_manager.fetch_job(queues=None)
```
Comment (ewjoachim) [inline tests/integration/test_manager.py:250]: I think this test was actually meaningless, even before your PR. We should have a look at it, in another PR.

File: tests/migration/test_migration.py [hunk 4]
```diff
@@ -13,7 +13,8 @@
 @pytest.fixture
 def run_migrations(db_execute):
     def _(dbname):
-        migrations = sorted(pathlib.Path("procrastinate/sql/migrations").glob("*.sql"))
+        folder = pathlib.Path(__file__).parents[2] / "procrastinate/sql/migrations"
```
Comment (ewjoachim) [inline tests/migration/test_migration.py:16]: So just to have it explicitely said, this allows the tests to be launched from a subfolder.

File: tests/unit/test_worker.py [hunk 5]
```diff
@@ -68,6 +68,54 @@ async def coro(*args, **kwargs):
...
+
+    test_worker = worker.Worker(app, delete_jobs=delete_jobs)
+    test_worker.run_job = mocker.Mock(side_effect=side_effect or coro)
+    job = job_factory(id=1)
+    await test_worker.job_manager.defer_job_async(job)
+
+    await test_worker.process_job(job=job)
+
+    assert 1 in connector.jobs
+
```
Comment (ewjoachim) [inline tests/unit/test_worker.py:117]: Those 2 new tests are A-OK !

**Review by ewjoachim — APPROVED** [review]
This looks amazing :) Thank you so much for your contribution !

**Comment by elemoine** [comment]
This looks really good, I just want to make sure that we do the right thing w.r.t. migrations. I'd like to have a closer look at that part specifically before we merge.

File: procrastinate/sql/schema.sql [hunk 6]
```diff
@@ -198,9 +199,25 @@ BEGIN
 END;
 $$;
 
+-- procrastinate_finish_job
+CREATE FUNCTION procrastinate_finish_job(job_id integer, end_status procrastinate_job_status, delete_job boolean) RETURNS void
+    LANGUAGE plpgsql
+AS $$
+BEGIN
+    IF delete_job THEN
+        DELETE FROM procrastinate_jobs where id = job_id;
```
Comment (elemoine) [inline procrastinate/sql/schema.sql:208]: (nit) s/where/WHERE

File: procrastinate/sql/migrations/00.17.00_02_delete_finished_jobs.sql [hunk 7]
```diff
@@ -0,0 +1,14 @@
+CREATE FUNCTION procrastinate_finish_job(job_id integer, end_status procrastinate_job_status, delete_job boolean) RETURNS void
+    LANGUAGE plpgsql
+AS $$
+BEGIN
+    IF delete_job THEN
+        DELETE FROM procrastinate_jobs where id = job_id;
```
Comment (elemoine) [inline procrastinate/sql/migrations/00.17.00_02_delete_finished_jobs.sql:6]: (nit) s/where/WHERE

**Review by elemoine — APPROVED** [review]
I made two minor comments that you may want to take into account before merging. Otherwise looks great to me. I paid attention to the migration part, and that looks good to me too! Merge away!

**Pull request merged**